packages/sources/token-balance/src/transport/evm.ts

```diff
@@ -0,0 +1,244 @@
+        continue
+      }
+
+      // extract network name from XXX_CHAIN_ID & get RPC_URL
+      const networkName = key.split(_CHAIN_ID)[0]
+      this.chainIdToStandardizedNetworkMap.set(chainId, networkName)
+
+      const rpcEnvVar = `${networkName}_RPC_URL`
+      const rpcUrl = process.env[rpcEnvVar]
+
+      if (!rpcUrl) {
+        logger.warn(`Missing RPC_URL for ${networkName}`)
+        continue
+      }
+
+      const provider = new ethers.JsonRpcProvider(rpcUrl, Number(chainId))
+      this.standardizedNetworkToProviderMap.set(networkName, provider)
+      logger.info(`created provider for network: ${networkName}, chain ID: ${chainId}`)
+    }
+  }
+
+  async initialize(
+    dependencies: TransportDependencies<BaseEndpointTypes>,
+    adapterSettings: BaseEndpointTypes['Settings'],
+    endpointName: string,
+    transportName: string,
+  ): Promise<void> {
+    await super.initialize(dependencies, adapterSettings, endpointName, transportName)
+    this.constructChainIdRpcMap()
+  }
+
+  async backgroundHandler(context: EndpointContext<BaseEndpointTypes>, entries: RequestParams[]) {
+    await Promise.all(entries.map(async (param) => this.handleRequest(param)))
+    await sleep(context.adapterSettings.BACKGROUND_EXECUTE_MS)
+  }
+
+  async handleRequest(param: RequestParams) {
+    let response: AdapterResponse<BaseEndpointTypes['Response']>
+    try {
+      response = await this._handleRequest(param)
+    } catch (e: unknown) {
+      const errorMessage = e instanceof Error ? e.message : 'Unknown error occurred'
+      logger.error(e, errorMessage)
+      response = {
+        statusCode: (e as AdapterInputError)?.statusCode || 502,
+        errorMessage,
+        timestamps: {
+          providerDataRequestedUnixMs: 0,
+          providerDataReceivedUnixMs: 0,
+          providerIndicatedTimeUnixMs: undefined,
+        },
+      }
+    }
+    await this.responseCache.write(this.name, [{ params: param, response }])
+  }
+
+  async _handleRequest(
+    param: RequestParams,
+  ): Promise<AdapterResponse<BaseEndpointTypes['Response']>> {
+    const { addresses } = param
+
+    // verify providers exist for this request
+    const normalizedAddresses = this.verifyProviders(addresses)
+
+    // construct list of decimals RPC requests
+    const constructDecimalsKey = (network: string, contractAddress: string): string =>
+      `${network}_${contractAddress}`
+
+    const decimalsMap = new Map<string, number | null>()
+    const decimalsRequests = []
+    for (const address of normalizedAddresses) {
+      const { network, contractAddress, decimalsSignature, provider } = address
+
+      // since we can have multiple addresses with balances of the same token,
+      // skip if we've encountered this pair already
+      const decimalsKey = constructDecimalsKey(network, contractAddress)
+      if (decimalsMap.has(decimalsKey)) {
+        logger.debug(`skipping decimals fetch for contract address ${address}`)
+        continue
+      }
```

Comment:
Is this needed?

Reply:
no, was a leftover from some assumptions before. Since we have the wallet [] and don't seem to be concerned about RPC usage credits I can remove
Only thing this would save is 1 decimals RPC request per address defined with a dupe contract/chain

Reply:
I miss understood something here, you can keep this.